Orange/canvas/canvas/items/linkitem.py

```diff
@@ -533,10 +538,35 @@ def __updateText(self):
         else:
             text = ""
 
-        self.linkTextItem.setHtml('<div align="center">{0}</div>'.format(text))
-        self.linkTextItem.document().setTextWidth(100)
-
+        self.linkTextItem.setHtml('<div align="center">{0}</div>'
+                                  .format(escape(text)))
         path = self.curveItem.curvePath()
+
+        # Constrain the text width if it is too long to fit on a single line
+        # between the two ends
+        if not path.isEmpty():
+            # Use the distance between the start/end points as a measure of
+            # available space
+            diff = path.pointAtPercent(0.0) - path.pointAtPercent(1.0)
+            available_width = math.sqrt(diff.x() ** 2 + diff.y() ** 2)
+        else:
+            available_width = -1
+
+        if available_width > 0:
+            # Get the ideal text width if it was unconstrained
+            doc = self.linkTextItem.document().clone(self)
+            doc.setTextWidth(-1)
+            idealwidth = doc.idealWidth()
+            doc.deleteLater()
+
+            # Constrain the text width  but not below a certain min width
+            minwidth = 100
+            textwidth = max(minwidth, min(available_width, idealwidth))
```

Comment:
Try running orange with:
```
QT_DRAW_SCENE_ITEM_RECTS=1 orange-canvas
```
to see the difference (the text label's bounding box would grow to the width/length of the entire link)